RAD-34 Create demo Android Application that illustrates MQTT

patience-mpofu opened the pull request:
This change creates custom UIs for devices to publish and subscribe to topics

Commit:
RAD-26 Create publish and subscribe UIs
This change creates custom UIs for devices to publish and subscribe to topics

Commit:
RAD-26 Correct subscribe UI
This change corrects the subscibe UI to match the given spec

Commit:
RAD-40 Refactor code
Remove all unnecessary code from the project
Remove all other sample projects
Create a new project and move all other code here

Commit:
RAD-41 Refactor subscription logic
Display text showing connection details and attempt to connect to new server

Commit:
RAD-41 Update subscribe screen
Updated the logic to subscribe using hardcoded literals for the connection
Removed unused code

Commit:
Add MqttSErvice to manifest
This change adds the service to the Android manifest and defines permissions that
the app should use. It also adds missing dependencies that were causing the app to crash

timler committed:
RAD-43: Refactor gradle.build to use proj props

Refactored the gradle build file to use the project properties to specify
the Android SDK version instead of it being hardcoded.

timler committed:
RAD-43: Add screen to view messages

Added screen to view messages using a Fragment and the RecycleView which
is a ListView that separates the view from the data.

Edited the main activity (activity_subscribe) to display messages as the
initial view because the content displaying the connection information
will now be displayed in settings.

Note: no work has been done yet on receiving a message. That will be
tackled next. This is just UI work.

https://www.sitepoint.com/mastering-complex-lists-with-the-android-recyclerview/

timler committed:
RAD-43: Use Android Log rather than System.out/err

Use Android Log rather than System.out/err so that errors will be captured
correctly.

Also renamed "addToHistory" method which just displayed a Toast.

timler committed:
RAD-43: Change formatting

Changed formatting in the layout XML so that it's easier to read/edit.

timler committed:
RAD-43: Fix disappearing app title

Fixed disappearing app title which was a problem when the number of messages
scrolled off the page.

https://stackoverflow.com/questions/32900559/collapsingtoolbarlayout-title-disappears

timler committed:
RAD-43: Change code formatting

Changed code formatting - part of 18cf43d - some lines were missed.

timler committed:
RAD-43: Remove nested scroll

Removed NestedScrollingView from RecyclerView because it was causing the
RecyclerView not to expand and scroll properly and it appears as if it
actually isn't needed.

timler committed:
RAD-43: Improvements to receiveing a message

Improved SubscribeActivity and MessageViewAdapter so that only one message
receiver callback was implemented (to avoid confusion) and that the
message is displayed after it is added to the screen.

Commit:
RAD-40 Add settings screen
Hard coded the connection settings on the settings screen

Commit:
Merge branch '34-demo-mqtt-app' of https://github.com/jembi/rad-mqtt-demo into 34-demo-mqtt-app

timler committed:
RAD-43: Make qos, cleanSession, etc constants

Made qos (quality of service), cleanSession and automaticReconnect constants
so that it's easy to toggle them and perhaps to add them to the properties.

timler committed:
RAD-43: Remove erroneous log message

timler committed:
RAD-40: Delete unused layout XML

timler committed:
RAD-44: Reformat XML file

timler committed:
RAD-44: Add icon to show if the app is offline

Added icon to the toolbar (menu) that shows if the app is offline - i.e.
not connected to the broker. This is handy because if there is a connection
problem, it can go offline permanently.

timler committed:
RAD-43: Replace fill_parent with match_parent

Replaced fill_parent with match_parent because fill_parent has been
deprecated.

Commit:
RAD-40 Make preferences editable
This change allows a user to edit the connection details

Commit:
Merge branch '34-demo-mqtt-app' of https://github.com/jembi/rad-mqtt-demo into 34-demo-mqtt-app

Commit:
RAD-40 Remove unused preferences

Commit:
RAD-40 Load connection details from preferences
This change allows dynamic assignment of the server connection details by loading them from preferences

Commit:
RAD-40 Removed unused strings

timler committed:
RAD-47: Set clientId to UUID if not set already

Default the clientId to a UUID string to avoid conflicts that causes
connection issues.

timler committed:
RAD-40: Remove unused strings

Commit:
RAD-40 General refactor
This change replaces library versions with variables for easier maintenance. It also removes
an unnecessary comment tag

Commit:
Merge branch '34-demo-mqtt-app' of https://github.com/jembi/rad-mqtt-demo into 34-demo-mqtt-app

timler committed:
RAD-45: Create MessageService and notification

Created a MessageService (to run the MQTT in the background) and display
Android notifications if the client receives a message.

There are still two issues (at least):
1. The notification icon uses the app icon, not the one specified
2. The background service doesn't run on its own thread so if the app
is stopped, it won't receive messages

https://developer.android.com/reference/android/app/Service.html
https://developer.android.com/training/notify-user/expanded.html
https://developer.android.com/reference/android/support/v4/app/NotificationCompat.Builder.html
http://www.vogella.com/tutorials/AndroidServices/article.html
https://developer.android.com/guide/topics/manifest/service-element.html#directBootAware
https://stackoverflow.com/questions/34863038/showing-a-snackbar-from-inside-a-service

Commit:
RAD-46 Implement database store
This change implements a database store to persist incoming messages

Commit:
RAD-46 Implement async database store
This change allows saving and retrieving data from the database asynchronously. This is necessary
to prevent the database calls from blocking the main UI thread

Commit:
RAD-46 Remove incorrect incorrect database close

timler committed:
RAD-47: Add static imports

Added static imports for a couple classes where it didn't make the code
more difficult to comprehend.

Commit:
RAD-46 Move private methods to classes that are using them
This changes moves the private methods into the classes that are using them to improve readability

Commit:
Merge branch '34-demo-mqtt-app' of https://github.com/jembi/rad-mqtt-demo into 34-demo-mqtt-app

timler committed:
RAD-48: Create app launcher

Created app launcher (overriding the default green Android one) - this
was important for Huawei phones which actually use the app icon for notification/
status icons.

timler committed:
Merge branch '34-demo-mqtt-app' of https://github.com/jembi/rad-mqtt-demo into 34-demo-mqtt-app

timler committed:
RAD-45: Remove comment

Removed comment that was incorrect to avoid confusion in the future.

timler committed:
RAD-45: Fix offline icon after receiving a msg

Fixed the offline icon so that it's by default off and it's only switched
on when the MQTT client notifies the app that it is offline.

timler committed:
Revert "RAD-45: Fix offline icon after receiving a msg"

This reverts commit f6e7dafc1c7bc780c3f0b848dc1e7495209e3197.

Commit:
RAD-46 Add functionality to retrieve older messages
This change implements functionality for the user to be able to see previously received messages

Commit:
Merge branch '34-demo-mqtt-app' of https://github.com/jembi/rad-mqtt-demo into 34-demo-mqtt-app

timler committed:
RAD-45: Improve offline icon hide/show

Improved the offline icon hide/show functionality so that if the app
is relaunched after receiving a message, it will display the correct
status.

Also changed from using getView to findItem which is more specific for
menus.

Allow for situations where the MQTT client updates the connection status
before the menu has been created.

timler committed:
RAD-49: Add uncaught exception handler

Added an uncaught exception handler so it's possible to see the logs for
exceptions that are thrown. Note: this is not a production ready solution.

timler committed:
RAD-45: Fix commented line of code

Fixed the commented line of code which was commented out for testing
and should not have been committed.

timler committed:
RAD-45: Upgrade paho MQTT client to 1.1.1

Upgraded paho MQTT client to 1.1.1 (from 1.1.0) due to a NPE being thrown
from the reconnect method.

To test:
1. Turn off mobile data & wifi
2. Stop the app
3. Start the app
4. Turn on wifi

You should see the offline icon change from being visible to invisible
and a notification about being connected.

https://github.com/eclipse/paho.mqtt.android/issues/220

Commit:
RAD-46 Change display order of messages
This change displays old messages oldest first to make it easier to find the most recent messages

Commit:
Merge branch '34-demo-mqtt-app' of https://github.com/jembi/rad-mqtt-demo into 34-demo-mqtt-app

Commit:
RAD-46 Remove System log statement since
we are now using LogCat

timler committed:
RAD-43: Remove boolean from activity subscribe

Removed boolean from activity subscribe which causes the collapsable
toolbar to lose the application name on occasion.

timler committed:
RAD-45: Disable notification if app is running

Disable the message notification if the app is running.

timler commented on org.jembi.rad.mqttdemo/src/main/java/org/jembi/rad/mqttdemo/database/MessageDBContract.java:
```diff
@@ -0,0 +1,28 @@
+
+/**
+ * Created by Jembi Health Systems on 2017/09/13.
+ */
```
After seeing this default Intellij comment, I was wondering if we should spend some time quickly documenting our code. Since this is a demo, and our first android app, it might be worth the effort.

timler commented on org.jembi.rad.mqttdemo/src/main/java/org/jembi/rad/mqttdemo/database/MessageDBOpenHelper.java:
```diff
@@ -0,0 +1,99 @@
+        String[] selectionArgs = {String.valueOf(new Date().getTime())};
+        String sortOrder = MessageEntry.COLUMN_NAME_DATE + " DESC";
+        open();
+        Cursor cursor = database.query(MessageEntry.TABLE_NAME,
```
the cursor needs to be closed

```
	    Cursor c = null;
		try {
		    c = db.rawQuery(query, values);
		    c.moveToFirst();
		     // do the reading of data
		} finally {
		    if (c != null) {
		        c.close();
		    }
		}		
	}
```

timler replied:
@patience-mpofu don't forget this comment

timler commented on org.jembi.rad.mqttdemo/src/main/java/org/jembi/rad/mqttdemo/database/MessageDBOpenHelper.java:
```diff
@@ -0,0 +1,99 @@
+    private List<Message> readFromCursor(Cursor cursor) {
+        List<Message> messages = new ArrayList<>();
+        while (cursor.moveToNext()) {
+            Long date = cursor.getLong(cursor.getColumnIndexOrThrow(COLUMN_NAME_DATE));
```
If you are throwing an exception here, you should add it to the method signature of 'readFromCursor' to be explicit to the callers of this method

timler replied:
@patience-mpofu also this comment

timler commented on org.jembi.rad.mqttdemo/src/main/java/org/jembi/rad/mqttdemo/database/MessageDBOpenHelper.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    public void insertMessage( Message message) {
+        open();
```
I wonder if it's necessary to call open here if it's done in the class constructor too? In theory it's more efficient to only open it when it's used, but since we know the application looks for messages immediately, there's also no harm in doing it in the constructor once.

timler commented on org.jembi.rad.mqttdemo/src/main/java/org/jembi/rad/mqttdemo/database/MessageDBOpenHelper.java:
```diff
@@ -0,0 +1,99 @@
+        String selectionCriteria = MessageEntry.COLUMN_NAME_DATE + " <  ?";
+        String[] selectionArgs = {String.valueOf(new Date().getTime())};
+        String sortOrder = MessageEntry.COLUMN_NAME_DATE + " DESC";
+        open();
```
See comment above

timler commented on org.jembi.rad.mqttdemo/src/main/java/org/jembi/rad/mqttdemo/database/MessageDBOpenHelper.java:
```diff
@@ -82,13 +106,43 @@ public void insertMessage( Message message) {
+        if(database != null) {
+            database.close();
+        }
+    }
```
I was reading that opening and closing db connections is so expensive that it's actually OK to leave the connection open for the running of the app and only close it in the onDestroy application callback

timler replied:
OK, I see you changed this in the next commit. Great 84f14ed

timler commented on org.jembi.rad.mqttdemo/src/main/java/org/jembi/rad/mqttdemo/SubscribeActivity.java:
```diff
@@ -44,8 +45,9 @@ protected void onCreate(Bundle savedInstanceState) {
         dbOpenHelper = new MessageDBOpenHelper(this);
-        if(!dbOpenHelper.getPreviousMessages().isEmpty()) {
-            for(Message message : dbOpenHelper.getPreviousMessages())  {
+        List<Message> previousMessages = dbOpenHelper.getPreviousMessages();
```
I am guessing this code has been changed in a future commit because it's not possible to immediately get a reply from an async operation... it'll need to be replaced with a callback

timler commented on org.jembi.rad.mqttdemo/src/main/java/org/jembi/rad/mqttdemo/MessageViewAdapter.java:
```diff
@@ -32,7 +32,6 @@ public void addMessage(Message message) {
         values.add(0, message);
         notifyItemInserted(0);
         Log.i("LOG", "Incoming message: " + message.getMessage());
-        System.out.println("Message index" + values.get(0).getMessage());
```
There's also a log in the onBind method that can be removed as it serves no helpful purpose. I added it, but it's now just annoying

timler commented on org.jembi.rad.mqttdemo/src/main/java/org/jembi/rad/mqttdemo/database/MessageDBOpenHelper.java:
```diff
@@ -54,7 +54,7 @@ public void onDowngrade(SQLiteDatabase sqLiteDatabase, int oldVersion, int newVe
     public List<Message> getPreviousMessages() {
         try {
-            new GetPreviousMessagesTask().execute();
+            return new GetPreviousMessagesTask().execute().get();
```
I'm a bit concerned this makes the task blocking. I was expecting to see a callback so that the messages are then added to the list (in SubscribeActivity) but I see you can call this method and expect a reply. I'll investigate a bit more, but I can't see how it wouldn't block the onCreate thread (of SubscribeActivity) until there is a suitable return

timler replied:
So yes, it definitely blocks the app. I added Thread.sleep(10000) into the doInBackground method of the GetPreviousMessagesTask and I saw a white screen until the timer was up

timler replied:
This is my idea at a solution.

1. Implement an interface that we can use as a callback

```
/**
 * Allows the Database to return data from an AsyncTask without blocking the thread.
 */
public interface DatabaseResult<T> {
    public void processResult(T result);
    public void processException(Exception e);
}
```

2. Change the database AsyncTask to accept an instance of this interface in its constructor, and call the processResult method in the onPostExecute method. For example:

```
    private class GetPreviousMessagesTask extends AsyncTask<Void, Void, List<Message>> {
        private DatabaseResult<List<Message>> callback;

        public GetPreviousMessagesTask(DatabaseResult<List<Message>> callback) {
            this.callback = callback;
        }

        @Override
        protected void onPreExecute() {
// snip
        }

        @Override
        protected List<Message> doInBackground(Void... params) {
// snip
        }

        @Override
        protected void onPostExecute(List<Message> messages) {
            callback.processResult(messages);
        }
// snip
    }
```

3. Change the code in SubscribeActivity to implement a DatabaseResult and add the messages to the UI in the callback method

```
        dbOpenHelper.getPreviousMessages(new DatabaseResult<List<Message>>() {
            @Override
            public void processResult(List<Message> previousMessages) {
                if (!previousMessages.isEmpty()) {
                    for (Message message : previousMessages) {
                        messageAdapter.addMessage(message);
                    }
                } else {
                    messageAdapter.addMessage(new Message(new Date(), "Welcome to the RAD MQTT Demo App"));
                }
            }

            @Override
            public void processException(Exception e) {
                // for now, do nothing - it's already been logged
            }
        });
```

timler commented on org.jembi.rad.mqttdemo/src/main/java/org/jembi/rad/mqttdemo/MessageViewAdapter.java:
```diff
@@ -29,9 +29,10 @@ public MessageViewAdapter(List<Message> items) {
+        values.add(0, message);
+        notifyItemInserted(0);
         Log.i("LOG", "Incoming message: " + message.getMessage());
+        System.out.println("Message index" + values.get(0).getMessage());
```
Note: this requires the messages to be added in the correct order to the list. In other words, if the results aren't ordered from the database, they will not be inserted in the correct order. I'm not sure what to suggest. One idea is perhaps making this list ordered, but then we need to be careful about the calls to "notifyItemInserted" and "smoothScrollToPosition" (in SubscribeActivity).

timler committed:
RAD-50: Implement system wide log tag

Implemented a system wide log tag so that the tag can be changed easily
and can also be standardised across the app.

timler committed:
RAD-52: Implement SortedList in Message View

Implemented SortedList in the MessageViewAdapter to ensure that the messages
retain a sort based on the date they were received (latest message at the
top).

https://stackoverflow.com/questions/29795299/what-is-the-sortedlistt-working-with-recyclerview-adapter

timler committed:
RAD-42: Add javadoc and comments

Added Javadoc and comments to classes to make the demo more helpful to
others (and ourselves).

timler committed:
RAD-42: Remove github files used for the paho repo

timler committed:
RAD-42: Add Javadoc to RadMQTTDemoApplication

timler committed:
RAD-42: Remove .project and add to gitignore

Commit:
RAD-46 Implement database callback
This change implements a callback for retrieving previous messages to prevent the process from
blocking the main UI thread

timler committed:
RAD-42: Improve project README

timler committed:
RAD-42: Add Javadoc to database package

timler commented on org.jembi.rad.mqttdemo/src/main/java/org/jembi/rad/mqttdemo/database/MessageDBOpenHelper.java:
```diff
@@ -53,9 +54,9 @@ public void onDowngrade(SQLiteDatabase sqLiteDatabase, int oldVersion, int newVe
     }
 
-    public List<Message> getPreviousMessages() {
+    public List<Message> getPreviousMessages(DatabaseResult<List<Message>> databaseResult) {
```
This method should return void otherwise someone using the message could innocently expect a useful result to be returned and be confused when they always get an empty ArrayList

timler commented on org.jembi.rad.mqttdemo/src/main/java/org/jembi/rad/mqttdemo/database/MessageDBOpenHelper.java:
```diff
@@ -53,9 +54,9 @@ public void onDowngrade(SQLiteDatabase sqLiteDatabase, int oldVersion, int newVe
-            return new GetPreviousMessagesTask().execute().get();
+             new GetPreviousMessagesTask(databaseResult).execute();
         } catch (Exception e) {
             Log.e(RadMQTTDemoApplication.LOG_TAG, "Could not retrieve older messages due to error " + e.getMessage());
```
Probably should call the processException method of DatabaseResult

timler replied:
I looked at this again and I think this try/catch block should go into the AsyncTask method `doInBackground`

timler commented on org.jembi.rad.mqttdemo/src/main/java/org/jembi/rad/mqttdemo/database/MessageDBOpenHelper.java:
```diff
@@ -68,8 +69,12 @@ public void insertMessage(Message message) {
+            this.callback = callback;
+        }
 
-        @Override
```
I think this may have been deleted by accident?

timler commented on org.jembi.rad.mqttdemo/src/main/java/org/jembi/rad/mqttdemo/SubscribeActivity.java:
```diff
@@ -58,14 +59,25 @@ protected void onCreate(Bundle savedInstanceState) {
+
+            @Override
+            public void processException(Exception e) {
+                // for now, do nothing - it's already been logged
```
I know this was my code ...  but maybe we should display an alert (snackbar msg)?

Commit:
RAD-51 Create package for ui classes
This change moves the activity and adapter classes into the ui package to make the demo clean

timler committed:
RAD-42: Change in README

timler committed:
RAD-53: Remove Eclipse PAHO code from repo

Removed the Eclipse PAHO code from the repo because it can be referenced
as a regular dependency.

timler committed:
RAD-53: Remove paho IML and reference in gradle

timler committed:
RAD-42: Remove Eclipse license information

timler committed:
RAD-42: Remove about.html (an Eclipse file)

timler committed:
RAD-42: Remove notice.html (Eclipse specific file)

timler committed:
RAD-38: Fix Gradle build and update Travis yml

Fixed Gradle build so that it's possible to run gradle assemble on the
main project and run the build of the app (sub project).

Updated travis.yml to reference the correct app name (not the paho
sample apps) and also to add some email notification if the build fails.

Commit:
RAD-46 Refactor Async callback
This change refactors the Async callback for database retrieval to make it more understandable
1. The callback returns a void return type to prevent confusion should an empty list be returned.
2. The cursor is closed after use to prevent leaks elsewhere

Commit:
Merge branch '34-demo-mqtt-app' of https://github.com/jembi/rad-mqtt-demo into 34-demo-mqtt-app

timler commented on org.jembi.rad.mqttdemo/src/main/java/org/jembi/rad/mqttdemo/database/MessageDBOpenHelper.java:
```diff
@@ -119,6 +114,7 @@ protected void onPostExecute(List<Message> messages) {
                 String message = cursor.getString(cursor.getColumnIndexOrThrow(COLUMN_NAME_MESSAGE_TEXT));
                 messages.add(new Message(new Date(date), message));
             }
+            cursor.close();
```
this will need to be done in a finally block so that we can guarantee if will happen

timler committed:
RAD-46: Close cursor in a finally block

Closed the db cursor in a finally block to ensure that it always gets called
even if an exception is thrown before the close is reached.

timler committed:
RAD-38: Update build-tools version in travis.yml

Updated build-tools version in travis.yml because Travis was reporting a
license issue:

>Checking the license for package Android SDK Build-Tools 25 in /usr/local/android-sdk/licenses
>Warning: License for package Android SDK Build-Tools 25 not accepted.

timler committed:
RAD-38: Add android license manually

Added license for Android to avoid Travis errors in the build. Also reset
versions to the versions in build.gradle (which is most correct).

Commit:
RAD-34 Add link to instructions on using the emulator

Commit:
Merge branch '34-demo-mqtt-app' of https://github.com/jembi/rad-mqtt-demo into 34-demo-mqtt-app

timler committed:
RAD-54: Added emulator instructions to README

timler committed:
RAD-54: Change the demo topic name

Changed the demo topic name to conform to MQTT standards. Also the topic
that was previously used had some retained messages which were always being
sent when the app started, making it look broken.

http://www.hivemq.com/blog/mqtt-essentials-part-5-mqtt-topics-best-practices

timler committed:
RAD-54: Add example MQTT.fx script

Added example MQTT.fx script that can be used to automate a demo with
the app.

timler committed:
RAD-55: Fix bug in app when a new message arrives

Fixed bug in the app when a new message arrived which was generally thrown
when the app was first installed.

The notifyItemInserted was being called unnecessarily because an appropriate
method is being called in the SortedList.Callback.

Commit:
RAD-56 Dont display or save duplicate messages
This change prevents duplicate messages from being sent or saved

Commit:
Merge branch '34-demo-mqtt-app' of https://github.com/jembi/rad-mqtt-demo into 34-demo-mqtt-app

Commit:
Revert "RAD-56 Dont display or save duplicate messages"

This reverts commit f544b47313d4b6745bf9ccfc478b4ceb3510fbfb. The approach used
to resolve the issue was determined to be ineffective. A different approach
will be taken.

Commit:
RAD-56 Use database queries to prevent
saving or displaying duplicates

Commit:
RAD-56 Prevent display of duplicate messages
This change adds a callback to determine whether or not a message should be displayed. It
also removes the group by clause in the query to retrieve previous messages. This is no
longer necessary since duplicate messages will not be saved

timler committed:
RAD-56: Add note to troubleshooting section

Added note to troubleshooting section to explain about duplicate messages
being suppressed.

timler committed:
RAD-34: Change demo topic for the last time
